### PR TITLE
HDDS-8352. OM crash with NPE in OMKeyCommitRequest due to missing user info

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

- while set userInfo in create request in pre-execute, getUserIfNotExist() method is used to populate process userName for non-rpc call for HSync feature, where it can be triggered during OpenKeyCleanupService additionally.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8352

## How was this patch tested?

verified impact of existing test cases in integration test.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java [hunk 1]
```diff
@@ -105,7 +105,8 @@ public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
 
     return getOmRequest().toBuilder()
         .setCommitKeyRequest(commitKeyRequest.toBuilder()
-            .setKeyArgs(newKeyArgs)).setUserInfo(getUserInfo()).build();
+            .setKeyArgs(newKeyArgs))
+        .setUserInfo(getUserIfNotExists(ozoneManager)).build();
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:109]: The same `setUserInfo(getUserIfNotExists` logic is performed in `OMClientRequest#preExecute` (the base method in parent class).

https://github.com/apache/ozone/blob/40e6875a7e6b5be1d6bf617a441e1c9aab1eb72b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java#L98-L107

I wonder if `OMKeyCommitRequest` should call the parent method first thing, like some other requests:

https://github.com/apache/ozone/blob/40e6875a7e6b5be1d6bf617a441e1c9aab1eb72b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/security/OMGetDelegationTokenRequest.java#L64-L67

Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:109]: @adoroszlai Calling parent method will not work here as building of request is different, parent does same thing with dummy request building.

Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:109]: @sumitagrawl can you please explain what you mean by "dummy request building"?

Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:109]: @adoroszlai Its not dummy, its original request building with setting only userInfo and layout.
To support modification of request content to build by parent, a builder need to be exposed in parent class, which derived class can manipulate.

But still we are having 2 way to get UserInfo() ---

OMClientRequest {
    OMRequest.Builder builder = getOmRequest().toBuilder();
    public void preExecute(...) {
      builder
      .setUserInfo(getUserIfNotExists(ozoneManager))
      .build();
   }
}
DeriveClass extends OMClientRequest {
    public void preExecute(...) {
      // do modification
     builder.setCommitKeyRequest(...);
   }
}

Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:109]: @sumitagrawl `OMGetDelegationTokenRequest` and similar requests call `super.preExecute` at the start of the method to obtain the `OMRequest` updated by parent logic, then update with local changes as necessary.  It is not optimal, due to duplicate builder conversion, but it is correct.

Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java]: @adoroszlai Thanks for valuable input, I have updated the changes

**Review by adoroszlai — APPROVED** [review]
Thanks @sumitagrawl for updating the patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java [hunk 2]
```diff
@@ -81,7 +81,8 @@ public OMKeyCommitRequest(OMRequest omRequest, BucketLayout bucketLayout) {
 
   @Override
   public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
-    CommitKeyRequest commitKeyRequest = getOmRequest().getCommitKeyRequest();
```
Comment (krishnaasawa1) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java:84]: Changes looks straightforward   preExecute to invoke  -> preExecute of OMClientRequest to populate Userinfo and remove explicitly populating below via builder which was causing the issue.

**Review by krishnaasawa1 — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @sumitagrawl for the patch, @krishnaasawa1 for the review.

**Pull request merged**